### PR TITLE
Support https proxy servers

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY

This PR adds the ability to proxy over a proxy server that listens on https.

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

This will conflict with https://github.com/ansible/ansible/pull/83085

### CODE REVIEW

File: lib/ansible/module_utils/urls.py [hunk 1]
```diff
@@ -218,9 +220,109 @@ def __init__(self, message, import_traceback, module=None):
...
+                if self._tls_in_tls:
+                    self.sock = TLSinTLS(self.sock, self._context, self._tunnel_host)
+
+        def _tunnel(self):
+            """This is a noop, we don't want http.client.HTTPConnection._tunnel running before we have
+            a chance to have HTTPSConnection wrap the socket for HTTPS
+
+            The fact that http.client.HTTPConnection.connect establishes the tunnel is the 1st reason
+            HTTPS proxies aren't supported, the 2nd is that there is no native TLS in TLS, see TLSinTLS
+            """
```
Comment (sivel) [inline lib/ansible/module_utils/urls.py:311]: Overriding `_tunnel` to not do anything, and then calling `super()._tunnel()` are probably the biggest concerns with fragility and supportability here, but urllib3 does it too, kind of:  

https://github.com/urllib3/urllib3/blob/b7c2910a9f49ba0d58c8d4381500c208f9a24765/src/urllib3/connection.py#L242
https://github.com/urllib3/urllib3/blob/b7c2910a9f49ba0d58c8d4381500c208f9a24765/src/urllib3/connection.py#L631

File: lib/ansible/module_utils/urls.py [hunk 2]
```diff
@@ -218,9 +220,109 @@ def __init__(self, message, import_traceback, module=None):
...
+        def https_open(self, req):
+            selector = urlparse(req.selector)
+            tls_in_tls = True
+            if req.host == req._tunnel_host and selector.hostname:
+                # urllib is all confused, this is likely http over an https proxy
+                tls_in_tls = selector.scheme == 'https'
+                port = selector.port
+                if not port:
+                    port = 443 if tls_in_tls else 80
+                req._tunnel_host = f'{selector.hostname}:{port}'
```
Comment (sivel) [inline lib/ansible/module_utils/urls.py:323]: All of this weirdness is just related to the fact that the existing code never attempted this, but again we're poking a private attribute.

File: lib/ansible/module_utils/urls.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansible/module_utils/urls.py:311]: This doesn't look like it should be a docstring but rather a comment. Could you change?

File: test/integration/targets/uri/tasks/install-proxy-and-test.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+- when: ansible_facts.distribution not in ['MacOSX']
```
Comment (webknjaz) [inline test/integration/targets/uri/tasks/install-proxy-and-test.yml:1]: It's possible to also test a TLS proxy by combining https://pypi.org/p/trustme and https://pypi.org/p/proxy.py. This is what I integrated into the `aiohttp` test suite back when @jborean93, Satellite folks and I were poking on how to implement TLS-in-TLS transport with aiohttp a few years back. Well, the `trustme` was pre-integrated earlier to test true TLS w/o disabling trust verification.

I also integrated `trustme` into the ansible-core integration tests in December: https://github.com/ansible/ansible/pull/82334. So we do have usage examples in our codebase.

Please, consider using it here as well to make it possible to run this integration test against macOS too.

File: test/integration/targets/uri/tasks/proxy.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      StartServers 1
+      Allow 127.0.0.1
+      Allow ::1
+      Allow ::
+      ViaProxyName "tinyproxy"
+      # LogFile "/tmp/tinyproxy.log"
+
+- name: Generate cert for stunnel
+  command: >-
+    openssl req -new -x509 -days 365 -nodes -out stunnel.pem -keyout stunnel.pem \
```
Comment (webknjaz) [inline test/integration/targets/uri/tasks/proxy.yml:29]: Based on the low-level TLS management I see here I believe that my suggestion to use `trustme` might simplify the configuration and positively affect the maintainability: https://github.com/ansible/ansible/pull/83089/files#r1571440157.

File: lib/ansible/module_utils/urls.py [hunk 5]
```diff
@@ -218,9 +220,109 @@ def __init__(self, message, import_traceback, module=None):
...
+        self.socket = socket
+
+        self.sslobj = sslobj = context.wrap_bio(
+            self.incoming, self.outgoing, server_hostname=server_hostname
+        )
+
+        self.read = functools.partial(self._process, True, sslobj.read)
+        self.recv = functools.partial(self._process, True, sslobj.read)
+        self.send = functools.partial(self._process, False, sslobj.write)
+        self.sendall = functools.partial(self._process, False, sslobj.write)
```
Comment (jborean93) [inline lib/ansible/module_utils/urls.py:237]: Do we need to worry about something calling `sendall` for this when the payload exceeds 16KiB? Is that even something we can do, i.e. a POST request with a large file?

File: lib/ansible/module_utils/urls.py [hunk 6]
```diff
@@ -218,9 +220,109 @@ def __init__(self, message, import_traceback, module=None):
...
+        self._decref_socketios = socket._decref_socketios
```
Comment (jborean93) [inline lib/ansible/module_utils/urls.py:238]: Is something in the `http.client.HTTPSConnection` calling this internally?

**Review by jborean93 — COMMENTED** [review]
I'll have to try out a few scenarios, this seemed like a fun one.

File: lib/ansible/module_utils/urls.py (hunk 5 above)
Comment (jborean93) [reply lib/ansible/module_utils/urls.py:237]: Ah just realized this is for TLS in TLS so it should never be sending/receiving any packets larger than this anyway.

File: lib/ansible/module_utils/urls.py (hunk 6 above)
Comment (sivel) [reply lib/ansible/module_utils/urls.py:238]: `socket.SocketIO` used in `makefile` uses it in the `close` method, which is called by `http`.  I could possibly get around this by not using `socket.SocketIO` directly, or subclass it.

**Comment by sivel** [comment]
For the record, I'm not really in favor of merging this.  It's touching too much that I just don't feel is supportable. And maybe with a further investment in time and code, it could maybe be made a little better, but I'm leary of the potential burden on future me and/or other devs to support this.

I'm open to being wrong about this, and just looking for feedback.

**Comment by jborean93** [comment]
While I feel the chances of urllib/http.client changing the way things are called internally I definitely wouldn't want to be the person who has to debug this if something was to break. On the flip side I do see people having to use a TLS based proxy becoming a more common scenario as time goes on so this can be a useful feature to have. Unfortunately I don't see this really being fixed properly in CPython itself so I'm not sure what the best way forward would be.

**Comment by sivel** [comment]
Closing this per above.

**Pull request closed without merging**